File: symmetric-client-clib/src/db/sql/Row.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "db/sql/Row.h"

void SymRowEntry_destroy(SymRowEntry *this) {
    free(this);
}

SymRowEntry * SymRowEntry_new(char *value, int sqlType, int size) {
    SymRowEntry *entry = (SymRowEntry *) calloc(1, sizeof(SymRowEntry));
    entry->value = value;
    entry->sqlType = sqlType;
    entry->size = size;
    return entry;
}

void SymRow_put(SymRow *this, char *columnName, void *value, int sqlType, int size) {
    SymRowEntry *entry = SymRowEntry_new(value, sqlType, size);
    this->map->put(this->map, columnName, entry);
}

char * SymRow_getString(SymRow *this, char *columnName) {
    char *value = NULL;
    SymRowEntry *entry = (SymRowEntry *) this->map->get(this->map, columnName);
    if (entry != NULL) {
        value = entry->value;
    }
    return value;
}

char * SymRow_getStringNew(SymRow *this, char *columnName) {
    return SymStringBuilder_copy(this->getString(this, columnName));
}

int SymRow_getInt(SymRow *this, char *columnName) {
    int value = 0;
    char *str = SymRow_getString(this, columnName);
    if (str != NULL) {
        value = atoi(str);
    }
    return value;
}

unsigned short SymRow_getBoolean(SymRow *this, char *columnName) {
    unsigned short value = 0;
    char *str = SymRow_getString(this, columnName);
    if (str != NULL) {
        value = atoi(str);
    }
    return value;
}

long SymRow_getLong(SymRow *this, char *columnName) {
    long value = 0;
    char *str = SymRow_getString(this, columnName);
    if (str != NULL) {
        value = atol(str);
    }
    return value;
}

SymDate * SymRow_getDate(SymRow *this, char *columnName) {
    SymDate *date = NULL;
    char *str = SymRow_getString(this, columnName);
    if (str != NULL) {
        date = SymDate_new(str);
    }
    return date;
}

int SymRow_getSize(SymRow *this, char *columnName) {
    int value = 0;
    SymRowEntry *entry = (SymRowEntry *) this->map->get(this->map, columnName);
    if (entry != NULL) {
        value = entry->size;
    }
    return value;
}

int SymRow_getSqlType(SymRow *this, char *columnName) {
    int value = 0;
    SymRowEntry *entry = (SymRowEntry *) this->map->get(this->map, columnName);
    if (entry != NULL) {
        value = entry->sqlType;
    }
    return value;
}

char * SymRow_stringValue(SymRow *this) {
    char *value = NULL;
    SymStringArray *names = this->map->keys(this->map);
    if (names->size > 0) {
        value = SymStringBuilder_copy(this->map->get(this->map, names->get(names, 0)));
    }
    names->destroy(names);
    return value;
}

int SymRow_intValue(SymRow *this) {
    int value = 0;
    char *str = SymRow_stringValue(this);
    if (str != NULL) {
        value = atoi(str);
        free(str);
    }
    return value;
}

long SymRow_longValue(SymRow *this) {
    long value = 0;
    char *str = SymRow_stringValue(this);
    if (str != NULL) {
        value = atol(str);
        free(str);
    }
    return value;
}

unsigned short SymRow_booleanValue(SymRow *this) {
    unsigned short value = 0;
    char *str = SymRow_stringValue(this);
    if (str != NULL) {
        value = atoi(str);
        free(str);
    }
    return value;
}

SymDate * SymRow_dateValue(SymRow *this) {
    SymDate *value = NULL;
    char *str = SymRow_stringValue(this);
    if (str != NULL) {
        value = SymDate_newWithString(str);
        free(str);
    }
    return value;
}

void SymRow_destroy(SymRow *this) {
    this->map->destroyAll(this->map, (void *) &SymRowEntry_destroy);
    free(this);
}

SymRow * SymRow_new(SymRow *this, int columnCount) {
    if (this == NULL) {
        this = (SymRow *) calloc(1, sizeof(SymRow));
    }
    this->map = SymMap_new(NULL, columnCount);
    this->put = (void *) &SymRow_put;
    this->getInt = (void *) &SymRow_getInt;
    this->getLong = (void *) &SymRow_getLong;
    this->getString = (void *) &SymRow_getString;
    this->getStringNew = (void *) &SymRow_getStringNew;
    this->getBoolean = (void *) &SymRow_getBoolean;
    this->getDate = (void *) &SymRow_getDate;
    this->getSize = (void *) &SymRow_getSize;
    this->getSqlType = (void *) &SymRow_getSqlType;
    this->stringValue = (void *) &SymRow_stringValue;
    this->intValue = (void *) &SymRow_intValue;
    this->longValue = (void *) &SymRow_longValue;
    this->booleanValue = (void *) &SymRow_booleanValue;
    this->dateValue = (void *) &SymRow_dateValue;
    this->destroy = (void *) &SymRow_destroy;
    return this;
}

File: symmetric-db/src/main/java/org/jumpmind/db/sql/Row.java
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.jumpmind.db.sql;

import java.io.IOException;
import java.math.BigDecimal;
import java.sql.Blob;
import java.sql.Clob;
import java.sql.SQLException;
import java.sql.Time;
import java.sql.Timestamp;
import java.util.Collection;
import java.util.Date;
import java.util.Map;

import org.apache.commons.io.IOUtils;
import org.jumpmind.exception.IoException;
import org.jumpmind.exception.ParseException;
import org.jumpmind.util.FormatUtils;
import org.jumpmind.util.LinkedCaseInsensitiveMap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Row extends LinkedCaseInsensitiveMap<Object> {

    private static final long serialVersionUID = 1L;

    protected Logger log = LoggerFactory.getLogger(getClass());

    public Row(int numberOfColumns) {
        super(numberOfColumns);
    }

    public Row(String columnName, Object value) {
        super(1);
        put(columnName, value);
    }

    public byte[] bytesValue() {
        Object obj = this.values().iterator().next();
        return toBytes(obj);
    }

    protected byte[] toBytes(Object obj) {
        if (obj != null) {
            if (obj instanceof byte[]) {
                return (byte[]) obj;
            } else if (obj instanceof Blob) {
                Blob blob = (Blob) obj;
                try {
                    return IOUtils.toByteArray(blob.getBinaryStream());
                } catch (IOException e) {
                    throw new IoException(e);
                } catch (SQLException e) {
                    throw new SqlException(e);
                }
            } else if (obj instanceof String) {
                return obj.toString().getBytes();
            } else {
                throw new IllegalStateException(String.format(
                        "Cannot translate a %s into a byte[]", obj.getClass().getName()));
            }
        } else {
            return null;
        }
    }

    public Number numberValue() {
        Object obj = this.values().iterator().next();
        if (obj != null) {
            if (obj instanceof Number) {
                return (Number) obj;
            } else {
                return new BigDecimal(obj.toString());
            }
        } else {
            return null;
        }
    }

    public Date dateValue() {
        Object obj = this.values().iterator().next();
        if (obj != null) {
            if (obj instanceof Date) {
                return (Date) obj;
            } else {
                return Timestamp.valueOf(obj.toString());
            }
        } else {
            return null;
        }
    }
    
    public Long longValue() {
        Object obj = this.values().iterator().next();
        if (obj != null) {
            if (obj instanceof Long) {
                return (Long)obj;
            } else {
                return Long.valueOf(obj.toString());    
            }            
        } else {
            return null;
        }        
    }
    

    public String stringValue() {
        Object obj = this.values().iterator().next();
        if (obj != null) {
            return obj.toString();
        } else {
            return null;
        }
    }
    
    public String csvValue() {
        StringBuilder concatenatedRow = new StringBuilder();
        Collection<Object> objs = this.values();
        int index = 0;
        for (Object obj : objs) {
            if (index > 0) {
                concatenatedRow.append(",");
            }
            concatenatedRow.append(obj != null ? obj.toString() : "");
            index++;            
        }
        return concatenatedRow.toString(); 
    }

    public byte[] getBytes(String columnName) {
        Object obj = get(columnName);
        return toBytes(obj);
    }
    
    public String getString(String columnName) {
        return getString(columnName, true);
    }

    public String getString(String columnName, boolean checkForColumn) {
        Object obj = this.get(columnName);
        if (obj != null) {
            return obj.toString();
        } else {
            if (checkForColumn) {
                checkForColumn(columnName);
            }
            return null;
        }
    }

    public int getInt(String columnName) {
        Object obj = this.get(columnName);
        if (obj instanceof Number) {
            return ((Number) obj).intValue();
        } else if (obj instanceof String) {
            return Integer.parseInt(obj.toString());
        } else {
            checkForColumn(columnName);
            return 0;
        }
    }

    public long getLong(String columnName) {
        Object obj = this.get(columnName);
        if (obj instanceof Number) {
            return ((Number) obj).longValue();
        } else if (obj instanceof String) {
            return Long.parseLong(obj.toString());
        } else {
            checkForColumn(columnName);
            return 0;
        }
    }

    public float getFloat(String columnName) {
        Object obj = this.get(columnName);
        if (obj instanceof Number) {
            return ((Number) obj).floatValue();
        } else if (obj instanceof String) {
            return Float.parseFloat(obj.toString());
        } else {
            checkForColumn(columnName);
            return 0;
        }
    }

    public BigDecimal getBigDecimal(String columnName) {
        Object obj = this.get(columnName);
        if (obj instanceof BigDecimal) {
            return (BigDecimal) obj;
        } else if (obj instanceof String) {
            return new BigDecimal(obj.toString());
        } else if (obj instanceof Integer) {
            return new BigDecimal(((Integer)obj).intValue());
        } else {
            checkForColumn(columnName);
            return null;
        }
    }

    public boolean getBoolean(String columnName) {
        Object obj = this.get(columnName);
        if ("1".equals(obj)) {
            return true;
        } else if (obj instanceof Number) {
            int value = ((Number) obj).intValue();
            return value > 0 ? true : false;
        } else if (obj instanceof Boolean) {
            return (Boolean) obj;
        } else if (obj instanceof String) {
            return Boolean.parseBoolean((String) obj);
        } else {
            checkForColumn(columnName);
            return false;
        }
    }

    public Time getTime(String columnName) {
        Object obj = this.get(columnName);
        if (obj instanceof Time) {
            return (Time) obj;
        } else {
            Date date = getDateTime(columnName);
            return new Time(date.getTime());
        }
    }

    public Date getDateTime(String columnName) {
        Object obj = this.get(columnName);
        if (obj instanceof Number) {
            long value = ((Number) obj).longValue();
            return new Date(value);
        } else if (obj instanceof Date) {
            return (Date) obj;
        } else if (obj instanceof String) {
            try {
                return getDate((String) obj, FormatUtils.TIMESTAMP_PATTERNS);
            } catch (ParseException ex) {
                // on xerial sqlite jdbc dates come back as longs
                return new Date(Long.parseLong((String) obj));
            }
        } else {
            checkForColumn(columnName);
            return null;
        }
    }

    protected void checkForColumn(String columnName) {
        if (!containsKey(columnName)) {
            throw new ColumnNotFoundException(columnName);
        }
    }

    final private java.util.Date getDate(String value, String[] pattern) {
    	int spaceIndex = value.lastIndexOf(" ");
        int fractionIndex = value.lastIndexOf(".");
        if (spaceIndex > 0 && fractionIndex > 0 && value.substring(fractionIndex, value.length()).length() > 3) {
            return Timestamp.valueOf(value);
        } else {
            return FormatUtils.parseDate(value, pattern);
        }
    }
    
    public Object[] toArray(String[] keys) {
        Object[] values = new Object[keys.length];
        for (int i = 0; i < keys.length; i++) {
            values[i] = get(keys[i]);
        }
        return values;
    }
    
    public String[] toStringArray(String[] keys) {
        String[] values = new String[keys.length];
        for (int i = 0; i < keys.length; i++) {
            values[i] = getString(keys[i]);
        }
        return values;
    }

    public long getLength() {
        long length = 0;
        
        for (Map.Entry<String, Object> entry : this.entrySet()) {
            try {
                Object obj = entry.getValue();
                if (obj instanceof Blob) {
                    length += ((Blob) obj).length();
                } else if (obj instanceof Clob) {
                    length += ((Clob) obj).length();
                } else {
                    length += obj.toString().length();
                }
            } catch (SQLException se) {
                log.warn("Unable to determine length of row, failure on column " + entry.getKey(), se);
            }
        }
        return length;
    }
}
